Refactor Error Handling + Misc

EternalDeiwos opened the pull request:
### Breaking Changes

- Major database refactor, changing of table ids to use UUIDv7.
- Several entities now feature soft-deletion using a `deleted_at` column.

### Features

- Error handling using exception filters to increase application stability.
- Better success and error reporting for interactions
- Crew owners can be added to a specified role for addressing crew leadership as a group
- Crew owner role can be set per guild using a command
- Crew creation can be limited to a specified role
- Crew creators role can be set per guild using a command
- Crews can be configured as `secure only` using a command
- Crews configured as `secure only` can only have their status and tickets displayed in non-public channels
- Crew logs will show the display the name and avatar of the author in the footer
- Commands and event listeners added to remove crew members who are no longer able to access a crew
- A global log channel can be configured per guild using a command
- If a global log channel is configured then new crew logs will be copied into the specified channel
- New crews may optionally also create a voice channel of the same name
  - If no global voice category is configured then the voice channel is created in the same category as the crew
- A global voice category can be configured per guild using a command
- Tickets originating from a different guild will now indicate this.

### Fixes

- Reworked privilege checks for privileged commands, incl. `purge`, `archive`, and `move`.

EternalDeiwos commented on supabase/migrations/20240714205505_first_tables.sql:
```diff
@@ -0,0 +1,332 @@
+
+alter table "public"."items" enable row level security;
+
+create table "public"."location" (
```
Lets add some QoL here:

- `iso2` - 2 character abbreviation for location, nullable
- `iso3` - 3 character abbreviation for location, nullable
- `slang` - `text[]`, default `[]`, additional terms used to describe a location that do not directly relate to their actual name. e.g. Cpass = Calahan's Passage.

P0etC replied:
Done

EternalDeiwos commented on supabase/migrations/20240714205505_first_tables.sql:
```diff
@@ -0,0 +1,332 @@
+
+alter table "public"."location" enable row level security;
+
+create table "public"."stockpile" (
```
We need to add something to track which war a stockpile belongs to. I would like to do a separate table, so that we can track war start and war end as well, so that we can easily work out what is the current war.

P0etC replied:
You  think it  should there should be like a war table with war start and war end?

EternalDeiwos commented on supabase/migrations/20240714205505_first_tables.sql:
```diff
@@ -0,0 +1,332 @@
+    "id" integer generated by default as identity not null,
+    "location_id" bigint not null,
+    "name" text not null,
+    "code" text not null,
```
This should be nullable, or default to `000000`. I want to ignore this column until we actually plan to use it. Or we could just omit the column entirely.

P0etC replied:
Done

EternalDeiwos commented on supabase/migrations/20240714205505_first_tables.sql:
```diff
@@ -0,0 +1,332 @@
+    "guild_sf" bigint not null,
+    "created_by_sf" bigint not null,
+    "created_at" timestamp with time zone not null default now(),
+    "crated" boolean not null
```
This should be an integer. We'll keep two counts per row.

P0etC replied:
Done

EternalDeiwos commented on supabase/migrations/20240714205505_first_tables.sql:
```diff
@@ -0,0 +1,332 @@
+
+grant update on table "public"."items" to "anon";
+
+grant delete on table "public"."items" to "authenticated";
```
Remove any privileges from all roles except `service_role`. The bot will be using that role. We can add more later if necessary.

P0etC replied:
I looked into the docs and tried to find a place where I can do it but really cant find a place to change roles on mass most of the roles are even protected from  me  and I cant change them.

P0etC replied:
Added war table  like it was proposed in the last discord schema  and added war number to stockpile table

EternalDeiwos reviewed (changes requested):
Take a look at comments and let me know what you think @P0etC

P0etC commented:
I got inspired by your pr and now i dont use the auto migration tool and created  migration files for every table seperatly. This shouuld also  help   with review. @EternalDeiwos would be great if you could take  a look at it.

EternalDeiwos commented on supabase/migrations/20240805185042_stockpiles.sql:
```diff
@@ -0,0 +1,49 @@
+create table "public"."stockpile" (
```
Inconsistent capitalization on SQL keywords. Lets make all keywords upper case.

EternalDeiwos commented on supabase/migrations/20240805185042_stockpiles.sql:
```diff
@@ -0,0 +1,49 @@
+create table "public"."stockpile" (
+    "id" integer generated by default as identity not null,
+    "location_id" bigint not null,
```
Lets make this column `poi_id` to match new naming.

EternalDeiwos commented on supabase/migrations/20240805185042_stockpiles.sql:
```diff
@@ -0,0 +1,49 @@
+    "name" text not null,
+    "code" text not null default '00000'::text,
+    "guild_sf" bigint not null,
+    "role_sf" bigint not null,
```
Drop this column to match new design

EternalDeiwos reviewed (changes requested):
Some small changes requested inline. 

Major change is related to the `items` table — I've put together a PR #4 with a `catalog` table which includes logic to import the JSON file and a view that provides appropriate access to the fields. You'll need to review that and adjust this PR to replace the `items` table.

EternalDeiwos commented:
Oh and also we need to use `_` instead of `-` in table names because postgres does not like `-`.

EternalDeiwos commented:
In the interest of moving forward, I've implemented this with #5.

glinklater committed:
chore: basic upgrads and deploy rmq

glinklater committed:
breaking: migrate in new structure and entities

glinklater committed:
feat(tickets): sort order function

glinklater committed:
refactor(*/repository): extend common repository

glinklater committed:
refactor(core): depend on external bot module

glinklater committed:
refactor: remove OperationStatus [wip]

glinklater committed:
refactor: remove OperationStatus

glinklater committed:
refactor(tag): generate id using uuid-v7

glinklater committed:
refactor: several data model changes

glinklater committed:
refactor: use uuidv7 for region-log

glinklater committed:
refactor: finish new error handling refactor

glinklater committed:
feat(crew): improve access testing and security

https://discord.com/channels/711362115570761739/1279209947841691719
https://discord.com/channels/711362115570761739/1280276220101136417
https://discord.com/channels/711362115570761739/1280983082203086878

glinklater committed:
feat(crew/log): show log author in footer

glinklater committed:
chore: some cleanup

glinklater committed:
fix(crew): admin command changing permissions for incorrect user

glinklater committed:
feat(crew): listen and reconcile crew membership with discord roles

glinklater committed:
feat(ticket): identify guild of origin for tickets

glinklater committed:
feat(crew): optionally create voice channels for crews

glinklater committed:
fix(crew): voice channel creation

glinklater committed:
fix: all uuid columns to use uuid type

glinklater committed:
feat: one-way migration for mass refactor

glinklater committed:
fix(crew): correct guild resolution

glinklater committed:
fix(game): ensure uuidv7 ids

glinklater committed:
chore(bot): display stack trace on general error handling

glinklater committed:
feat(game): database migration for game tables
